### PR TITLE
Feature/fallback return

### PR DESCRIPTION
**Pull request by novaknole**
## Description

Please include a summary of the change and be sure you follow the contributions rules we do provide [here](./CONTRIBUTIONS.md)

Task: [ID]()

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: packages/contracts/contracts/test/CallbackHandlerHelperMock.sol [hunk 1]
```diff
@@ -2,17 +2,23 @@
...
     address callbackHandlerMockAddr;
 
-    constructor(address _callbackHandlerMockAddr) {
-        callbackHandlerMockAddr = _callbackHandlerMockAddr;
+    /// @notice Executes _handleCallback on the parens for testing purposes.
+    /// @param selector Any kind of selector in order to call fallback.
+    /// @param data Some arbitrary data. Mainly, you should pass msg.data to _handleCallback most of the time.
+    function handleCallback(bytes4 selector, bytes memory data) external returns (bytes4) {
+        bytes4 magicNumber = _handleCallback(selector, data);      
+        return magicNumber;
```
Comment (brickpop) [inline packages/contracts/contracts/test/CallbackHandlerHelperMock.sol:15]: Does the stack-saving practice apply here?

```suggestion
    function handleCallback(bytes4 selector, bytes memory data) external returns (bytes4 magicNumber) {
        magicNumber = _handleCallback(selector, data);      
```

Comment (novaknole) [reply packages/contracts/contracts/test/CallbackHandlerHelperMock.sol:15]: It's a test file, so I preferred to be more explicit.

File: packages/contracts/contracts/test/CallbackHandlerHelperMock.sol [hunk 2]
```diff
@@ -2,17 +2,23 @@
...
     }
 
-    // @notice Executes the `fallback()` function in `CallbackHandlerMock` and emits the returned value.
-    // @param selector Any kind of selector in order to call fallback.
-    function handleCallback(bytes4 selector) external returns (bytes32 magicNumber) {
-        (, bytes memory value) = callbackHandlerMockAddr.call(abi.encodeWithSelector(selector));
-        magicNumber = abi.decode(value, (bytes32));
+    /// @notice Executes _registerCallback on the parens to register magic number per selector.
```
Comment (mathewmeconry) [inline packages/contracts/contracts/test/CallbackHandlerHelperMock.sol:18]:
```suggestion
    /// @notice Executes `_registerCallback` on the parent to register magic number per selector.
```

File: packages/contracts/contracts/test/CallbackHandlerHelperMock.sol [hunk 3]
```diff
@@ -2,17 +2,23 @@
...
+    /// @param selector The function selector
```
Comment (mathewmeconry) [inline packages/contracts/contracts/test/CallbackHandlerHelperMock.sol:19]:
```suggestion
    /// @param selector The function selector.
```

**Review by mathewmeconry — CHANGES_REQUESTED** [review]

**Review by brickpop — APPROVED** [review]

File: packages/contracts/contracts/core/DAO.sol [hunk 4]
```diff
@@ -278,8 +278,11 @@ contract DAO is
     }
 
     /// @notice Fallback to handle future versions of the [ERC-165](https://eips.ethereum.org/EIPS/eip-165) standard.
-    fallback() external {
-        _handleCallback(msg.sig); // WARN: does a low-level return, any code below would be unreacheable
+    /// @param _input As per the solidity docs, _input is the alias for msg.data.
+    /// @return bytes The magic number registered on the signature selector.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/DAO.sol:282]:
```suggestion
    /// @param _input An alias being equivalent to `msg.data`. This a feature of the fallback function was introduced with [solidity compiler version 0.7.6](https://github.com/ethereum/solidity/releases/tag/v0.7.6)
    /// @return bytes The magic number registered for the function selector triggering the fallback.
```

File: packages/contracts/contracts/core/component/CallbackHandler.sol [hunk 5]
```diff
@@ -6,37 +6,45 @@ pragma solidity 0.8.10;
...
+    /// @param data The calldata for the function signature.
+    event CallbackReceived(bytes4 indexed sig, bytes data);
 
     /// @notice Handles callbacks to adaptively support ERC standards.
-    /// @param _callbackSelector The selector of the callback function.
     /// @dev This function is supposed to be called via `_handleCallback(msg.sig, msg.data)` in the `fallback()` function of the inheriting contract.
-    function _handleCallback(bytes4 _callbackSelector) internal view {
-        bytes32 magicNumber = callbackMagicNumbers[_callbackSelector];
+    /// @param _callbackSelector The selector of the callback function.
+    /// @return bytes4 The magic number stored for `_callbackSelector`.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/component/CallbackHandler.sol:29]:
```suggestion
    /// @param _callbackSelector The function selector of the callback function.
    /// @return The magic number registered for the function selector triggering the fallback.
```

File: packages/contracts/contracts/core/component/CallbackHandler.sol [hunk 6]
```diff
@@ -6,37 +6,45 @@ pragma solidity 0.8.10;
...
         if (magicNumber == UNREGISTERED_CALLBACK) {
             revert UnkownCallback({callbackSelector: _callbackSelector, magicNumber: magicNumber});
         }
-        // low-level return magic number
-        assembly {
-            mstore(0x00, magicNumber)
-            return(0x00, 0x20)
-        }
+
+        emit CallbackReceived(_callbackSelector, _data);
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/component/CallbackHandler.sol:39]:
```suggestion
        emit CallbackReceived({sig: _callbackSelector, data: _data});
```

File: packages/contracts/contracts/test/CallbackHandlerHelperMock.sol [hunk 7]
```diff
@@ -2,17 +2,23 @@
...
+import {CallbackHandler} from "../core/component/CallbackHandler.sol";
+
+contract CallbackHandlerMockHelper is CallbackHandler {
     address callbackHandlerMockAddr;
 
-    constructor(address _callbackHandlerMockAddr) {
-        callbackHandlerMockAddr = _callbackHandlerMockAddr;
+    /// @notice Executes _handleCallback on the parens for testing purposes.
+    /// @param selector Any kind of selector in order to call fallback.
+    /// @param data Some arbitrary data. Mainly, you should pass msg.data to _handleCallback most of the time.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/test/CallbackHandlerHelperMock.sol:12]:
```suggestion
    /// @notice Calls the internal `_handleCallback` on the parent `CallbackHandler` for testing purposes.
    /// @param selector The function selector of the callback function to be tested.
    /// @param data Arbitrary data accompanying the callback that will be emitted with the `CallbackReceived` event.
```

**Review by Michael-A-Heuer — CHANGES_REQUESTED** [review]
Looking good. Can be approved once NatSpec suggestions are merged.

File: packages/contracts/contracts/core/DAO.sol [hunk 8]
```diff
@@ -278,8 +278,11 @@ contract DAO is
     }
 
     /// @notice Fallback to handle future versions of the [ERC-165](https://eips.ethereum.org/EIPS/eip-165) standard.
-    fallback() external {
-        _handleCallback(msg.sig); // WARN: does a low-level return, any code below would be unreacheable
+    /// @param _input An alias being equivalent to `msg.data`. This a feature of the fallback function was introduced with [solidity compiler version 0.7.6](https://github.com/ethereum/solidity/releases/tag/v0.7.6)
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/DAO.sol:281]:
```suggestion
    /// @param _input An alias being equivalent to `msg.data`. This feature of the fallback function was introduced with [solidity compiler version 0.7.6](https://github.com/ethereum/solidity/releases/tag/v0.7.6)
```

File: packages/contracts/contracts/core/component/CallbackHandler.sol [hunk 9]
```diff
@@ -6,37 +6,45 @@ pragma solidity 0.8.10;
...
+    /// @param data The calldata for the function signature.
+    event CallbackReceived(bytes4 indexed sig, bytes data);
 
     /// @notice Handles callbacks to adaptively support ERC standards.
-    /// @param _callbackSelector The selector of the callback function.
     /// @dev This function is supposed to be called via `_handleCallback(msg.sig, msg.data)` in the `fallback()` function of the inheriting contract.
-    function _handleCallback(bytes4 _callbackSelector) internal view {
-        bytes32 magicNumber = callbackMagicNumbers[_callbackSelector];
+    /// @param _callbackSelector The function selector of the callback function.
+    /// @return The magic number registered for the function selector triggering the fallback.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/component/CallbackHandler.sol:29]:
```suggestion
    /// @return The magic number registered for the function selector triggering the callback.
```

**Review by Michael-A-Heuer — APPROVED** [review]

**Review by brickpop — APPROVED** [review]

**Pull request merged**